File: santander customer transaction prediction.R
rm(list = ls())

# Set timer for script run
st = list()
st[['Start']] = Sys.time()

options(max.print = 200000)

###############################################################################################################
###############################################################################################################

## load libraries
print("--- Loading Libraries ---")
library(caret)
library(dplyr)
library(usdm)
library(MLmetrics)
library(skimr)
library(ggplot2)
library(jpeg)
library(ROSE)
library(randomForest)
library(e1071)

###############################################################################################################
###############################################################################################################

## setting the working directory
setwd('c:/users/user/project')

###############################################################################################################
###############################################################################################################

## load dataset
train = read.csv('train.csv')
test = read.csv('test.csv')

## EDA
dim(train)
str(train)
class(train)
summary(train)

# Backing up the ID_code field
key.train = train$ID_code
key.test = test$ID_code

# Removing ID variables
train$ID_code = NULL
test$ID_code = NULL

###############################################################################################################
###############################################################################################################

# Check for missing values
anyNA(train)
print("No Missing values found")

################################## No missing values found ####################################################

## Preparing data for analysis
training = train

# Checking multicollinearity
vifcor(training)

#Bringing all features to a single scale
training[,2:201] = scale(training[,2:201])
test[,1:200] = scale(training[,1:200])

# Reformatting the levels of the target variable class
training$target = as.factor(training$target)
levels(training$target) = c('zero', 'one')

# Splitting dataframe into independent and dependent variables

independent_var = colnames(training) != 'target'
x = training[,independent_var]
y = training$target

#################### No variable from the 200 input variables has collinearity problem. #######################

# Statistical summary with histograms
skimmed = skim_to_wide(x)
print(skimmed)

# Create a generic function for generating distribution & box plots
buildplot =function(x, type = 'dist'){
  variablenames = colnames(x)
  temp = as.numeric(1)
  for(i in seq(8, dim(x)[2], 8)){
    par(mar = c(2, 2, 2, 2))
    par(mfrow = c(4, 2))
    for (j in variablenames[temp:i]){
      if(type == 'box'){
        jpeg(filename = paste('boxplot', j, '.jpg', sep = ' '), width = 1080, height = 1080)
        boxplot(x[[j]], main = j, col = 'grey' )
        dev.off()
      }
      
      else{
        jpeg(filename = paste('dplot', j, '.jpg', sep = ' '), width = 1080, height = 1080)
        plot(density(x[[j]]), main = j, col = 'red' )
        dev.off()
      }
    }
    temp = i + 1
  }
}
# Distribution plot for training & test data 
buildplot(x)
buildplot(test)

######### The distribution of the independant variables in both train & test datasets is fairly normal ########

# Box plot for outlier analysis of training & test data 
buildplot(x, 'box')
buildplot(test, 'box')

######### The distribution of the independant variables in both train & test datasets is fairly normal ########

###############################################################################################################
###############################################################################################################

## Partitioning the train dataset for validation

# Backing up the complete processed train dataset
training_complete = training

# create a list of 75% of the rows in the train dataset we can use for training
validation_index = createDataPartition(training$target, p=0.75, list=FALSE)

#select 20% of the data for validation
validation = training[-validation_index,]

# use the remaining 80% of data to training and testing the models
training = training[validation_index,]

# Creating a stratified sample with oversampled positive target response
training_over = ovun.sample(target ~ ., training, method = 'over')$data

###############################################################################################################
###############################################################################################################

## Evaluating algorithms

# Build models

#Logistic Regression
Print("Lostic Regression:- ")
#setting seed
set.seed(2)
#fit the model
fit.glm = glm(target ~., data = training, family = binomial(link="logit"))
#test GLM on validation subset
# print(fit.glm)
predictions.glm = predict(fit.glm, validation)
predictions.glm = as.factor(ifelse(predictions.glm < 0.5, 'zero', 'one'))
#calculate confusion matrix
cm = ConfusionMatrix(predictions.glm, validation$target)
glm = list()
#calculate scores
glm[['accuracy']] = Accuracy(predictions.glm, validation$target) * 100
#plot ROC curve
glm[['fnr']] = (cm['one','zero'] * 100)/(cm['one','zero']+cm['one', 'one'])
glm[['auc']] = roc.curve(predictions.glm, validation$target)$auc
glm[['ConfusionMatrix']] = cm
#print scores and confusion matrix
print(glm)

#Random Forest
print("Random Forest:- ")
#Setting seed
set.seed(2)
#fit the model
fit.rf = randomForest(target ~., data = training, ntree = 20, mtry = 10)

#test RandomForest on validation subset
# print(fit.xgb)
predictions.rf = predict(fit.rf, validation)
#calculate confusion matrix
cm = ConfusionMatrix(predictions.rf, validation$target)
rf = list()
#calculate scores
rf[['accuracy']] = Accuracy(predictions.rf, validation$target) * 100
#plot ROC Curve
rf[['auc']] = roc.curve(predictions.rf, validation$target)$auc
rf[['fnr']] = (cm['one','zero'] * 100)/(cm['one','zero']+cm['one', 'one'])
rf[['ConfusionMatrix']] = cm
print(rf)

#GaussianNB
print("GaussainNB:-")
#setting seed
set.seed(2)
#fit the model
fit.naivebayes = naiveBayes(target ~., data = training)
#test NaiveBayes on validation subset
# print(fit.naivebayes)
predictions.naivebayes = predict(fit.naivebayes, validation)
#calculate confusion matrix
cm = ConfusionMatrix(predictions.naivebayes, validation$target)
naivebayes = list()
#calculate scores
naivebayes[['accuracy']] = Accuracy(predictions.naivebayes, validation$target) * 100
#plot ROc curve
naivebayes[['auc']] = roc.curve(predictions.naivebayes, validation$target)$auc
naivebayes[['fnr']] = (cm['one','zero'] * 100)/(cm['one','zero']+cm['one', 'one'])
naivebayes[['ConfusionMatrix']] = cm
print(naivebayes)


###############################################################################################################
###############################################################################################################


#test NaiveBayes on validation subset
# print(fit.naivebayes)
predictions.naivebayes = predict(fit.naivebayes, validation)
cm = ConfusionMatrix(predictions.naivebayes, validation$target)
naivebayes = list()
naivebayes[['accuracy']] = Accuracy(predictions.naivebayes, validation$target) * 100
naivebayes[['auc']] = roc.curve(predictions.naivebayes, validation$target)$auc
naivebayes[['fnr']] = (cm['one','zero'] * 100)/(cm['one','zero']+cm['one', 'one'])
naivebayes[['ConfusionMatrix']] = cm

results = list('glm' = glm, 'rf' = rf, 'naivebayes' = naivebayes)

###############################################################################################################
###############################################################################################################

## Evaluating algorithms with oversampled training dataset

# Build models
set.seed(2)
over.glm = glm(target ~., data = training_over, family = binomial(link='logit'))

set.seed(2)
over.rf = randomForest(target ~., data = training_over, ntree = 20, mtry = 10)

set.seed(2)
over.naivebayes = naiveBayes(target ~., data = training_over)

###############################################################################################################
###############################################################################################################

#test GLM on validation subset
# print(fit.glm)
predictions.glm = predict(over.glm, validation)
predictions.glm = as.factor(ifelse(predictions.glm < 0.5, 'zero', 'one'))
cm = ConfusionMatrix(predictions.glm, validation$target)
glm = list()
glm[['accuracy']] = Accuracy(predictions.glm, validation$target) * 100
glm[['auc']] = roc.curve(predictions.glm, validation$target)$auc
glm[['fnr']] = (cm['one','zero'] * 100)/(cm['one','zero']+cm['one', 'one'])
glm[['ConfusionMatrix']] = cm

#test RandomForest on validation subset
# print(fit.xgb)
predictions.rf = predict(over.rf, validation)
cm = ConfusionMatrix(predictions.rf, validation$target)
rf = list()
rf[['accuracy']] = Accuracy(predictions.rf, validation$target) * 100
rf[['auc']] = roc.curve(predictions.rf, validation$target)$auc
rf[['fnr']] = (cm['one','zero'] * 100)/(cm['one','zero']+cm['one', 'one'])
rf[['ConfusionMatrix']] = cm

#test NaiveBayes on validation subset
# print(fit.naivebayes)
predictions.naivebayes = predict(over.naivebayes, validation)
cm = ConfusionMatrix(predictions.naivebayes, validation$target)
naivebayes = list()
naivebayes[['accuracy']] = Accuracy(predictions.naivebayes, validation$target) * 100
naivebayes[['auc']] = roc.curve(predictions.naivebayes, validation$target)$auc
naivebayes[['fnr']] = (cm['one','zero'] * 100)/(cm['one','zero']+cm['one', 'one'])
naivebayes[['ConfusionMatrix']] = cm

results_over = list('glm' = glm, 'rf' = rf, 'naivebayes' = naivebayes)

###############################################################################################################
###############################################################################################################

# Selecting NaiveBayes algorithm modeled using complete train data to predict the test data
# since it provides the best ratio of Accuracy and FNR
Print("---Selecting GaussainNB algorithm model as it provides the high accuracy---")

# Creating a stratified sample with oversampled positive target response
training_over = ovun.sample(target ~ ., training_complete, method = 'over')$data

set.seed(2)
over.naivebayes = naiveBayes(target ~., data = training_over)

#test NaiveBayes on validation subset
# print(fit.naivebayes)
predictions.naivebayes = predict(over.naivebayes, test)
test$target = predictions.naivebayes
levels(test$target) = c('0','1')

# Saving the final dataset
write.csv(test, 'finalsubmission - r.csv')

st[['End']] = Sys.time()

print(paste('Total Script Runtime: ', st$End - st$Start, sep = ''))

print("---End of the program---")

File: Santander customer Transaction Prediction.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


#import libraries
print("---loading libariaries---")

import warnings
warnings.filterwarnings("ignore")

import pandas as pd 
import numpy as np 

#Complementary error function
from scipy.special import erfc

#Random Sampling 
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler


#Feature Scaling, Extraction
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

#Plot
import matplotlib.pyplot as plt
import seaborn as sns

#Model Selection
from sklearn.model_selection import train_test_split

#Machine Learning Algorthims
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import LinearSVC
from sklearn.naive_bayes import GaussianNB

#Metrics
from sklearn.metrics import roc_auc_score, plot_roc_curve, classification_report, confusion_matrix


#load data
df_train=pd.read_csv("train.csv")
df_test=pd.read_csv("test.csv")

#shape of train and test data
df_train.shape
df_test.shape

df_train.info()

#checking types
df_train.dtypes

#observing data
df_train.head(5)
df_test.head(5)

#counting observations per target class
df_train.target.value_counts()

#plotting pie chart for target class
df_train['target'].value_counts().plot(kind='pie', figsize=(5,5))

#checking for missing values in train data
df_train.isna().sum().sum()

#checking for missing values in train data
df_test.isna().sum().sum()

#Summary of train data
df_train.describe()


#checking outliers using Chauvenet's criterion
def chauvenet(array):
    mean = array.mean()           # Mean of incoming array
    stdv = array.std()            # Standard deviation
    N = len(array)                # Lenght of incoming array
    criterion = 1.0/(2*N)         # Chauvenet's criterion
    d = abs(array-mean)/stdv      # Distance of a value to mean in stdv's
    prob = erfc(d)                # Area normal dist.    
    return prob < criterion       # Use boolean array outside this function

numerical_features=df_train.columns[2:]

#outliers in each variable in train data 
train_outliers = dict()
for col in [col for col in numerical_features]:
    train_outliers[col] = df_train[chauvenet(df_train[col].values)].shape[0]
train_outliers = pd.Series(train_outliers)

#Plotting outliers in a barplot
train_outliers.sort_values().plot(figsize=(14, 40), kind='barh').set_xlabel('Number of outliers');


#printing otliers percentage of train dataset
print('Total number of outliers in training set: {} ({:.2f}%)'.format(sum(train_outliers.values), (sum(train_outliers.values) / df_train.shape[0]) * 100))

#outliers in each variable in test data 
test_outliers = dict()
for col in [col for col in numerical_features]:
    test_outliers[col] = df_test[chauvenet(df_test[col].values)].shape[0]
test_outliers = pd.Series(test_outliers)

#plotting outliers in a barplot
test_outliers.sort_values().plot(figsize=(14, 40), kind='barh').set_xlabel('Number of outliers');

#printing outliers percentage in test dataset
print('Total number of outliers in testing set: {} ({:.2f}%)'.format(sum(test_outliers.values), (sum(test_outliers.values) / df_test.shape[0]) * 100))

#remove these outliers in train and test data
for col in numerical_features:
    train=df_train.loc[(~chauvenet(df_train[col].values))]
for col in numerical_features:
    test=df_test.loc[(~chauvenet(df_test[col].values))]    
    
#shape of train and test data after removal of outliers 
train.shape,test.shape

#describe train data after removal of outliers
train.describe()

#describe test data after removal of outliers
test.describe()

#hisograms are used to check distribution of data 
#draw histograms of numeric data in training set 
print("Distribution of Columns")
plt.figure(figsize=(40,200))
for i,col in enumerate(numerical_features):
    plt.subplot(50,4,i+1)
    plt.hist(train[col])
    plt.title(col)  

print("Almost all features follow fairly normal distribution ")


#defining confusion matrix
def cnf_mtx(ytest, ypred, filename):
    import seaborn as sns
    import matplotlib.pyplot as plt 
    from sklearn.metrics import confusion_matrix
    plt.clf()
    cm = confusion_matrix(ytest,ypred)
    print("----confusion matrix-----")
    print(cm)
    plt.title(filename.split(".")[0]) # title with fontsize 20

    plot = sns.heatmap(cm, annot=True, fmt='g')
    plt.xlabel('Predicted labels') # x-axis label with fontsize 15
    plt.ylabel('True labels') # y-axis label with fontsize 15
    plot.get_figure().savefig(filename)
    

if __name__ == "__main__":
    print("Processing to machine learning algorithms by over sampling method ")

    print(str(train['target'].value_counts())+"\n\nCreating the X features and Y target variables..\n")

    X_train = train.drop(columns=['ID_code', 'target'])
    y_train = train['target']
    X_val = test.drop(columns='ID_code')

    #Fixing Class Imbalanced
    print("Found the Class Imbalanced \n\nFixing the Class imbalanced by over sampling with SMOTE.\n")

    randsmp =  SMOTE()
    X_train, y_train = randsmp.fit_resample(X_train, y_train)

    print("Creating a pipline for fit and transform features into standard scale and reducing features..\n")

    steps = [("Standarad Scalar", StandardScaler(with_mean=True)),
             ("PCA", PCA(n_components=10))
            ]

    pipline = Pipeline(steps=steps)
    pXtrain = pipline.fit_transform(X_train)
    

    #Split dataset into train and size of 25% in dataset as test data.
    print("Split dataset into Train and Test \n")

    Xtrain, Xtest, ytrain, ytest = train_test_split(pXtrain, y_train, test_size=0.25, random_state=42 )

    print("Now, Training the Machine Learning Model with Xtrain, ytrain. and Validating the Model by Xtest, ytest. \n")

    
    print("Logistic Regression:- ")
    #Logistic Regression
    lr = LogisticRegression()
    lr.fit(Xtrain, ytrain) ## GENETRATED WEIGHTS 

    lr_ypred = lr.predict(Xtest) # FINDING TARGETS

    #print classification report
    print(classification_report(ytest, lr_ypred)+"\n\n")
    #Print Accuracy
    print("AUC Score:- %.3f \n\n" % roc_auc_score(ytest, lr_ypred))
    
    #print confusion matrix
    cnf_mtx(ytest, lr_ypred, "LR-Confusion Matrix.png")
    #plot ROC Curve
    plot_roc_curve(lr, Xtest, ytest)
    #axis labels
    plt.xlabel("False Positive Rate")
    plt.ylabel("True Positive Rate")
    #show plot
    plt.show()
    
    
    print("RandomForest Classifier:- ")
    #RandomForest Classifer
    rnclf = RandomForestClassifier(n_estimators=5)
    rnclf.fit(Xtrain, ytrain)
    rnclf_ypred = rnclf.predict(Xtest)
    
    #Print Classification Reprt
    print(classification_report(ytest, rnclf_ypred)+"\n\n")
    #Print Accuracy
    print("AUC Score:- %.3f \n\n" % roc_auc_score(ytest, rnclf_ypred))
    
    #Print Confusion Matrix
    cnf_mtx(ytest, rnclf_ypred, "Rnclf-ConfusionMatrix.png")
    #plot Roc Curve
    plot_roc_curve(rnclf, Xtest, ytest)
    # axis labels
    plt.xlabel("False Positive Rate")
    plt.ylabel("True Positive Rate")
    #show plot
    plt.show()
    
    
    print("LinearSVC Classifier:- ")
    #Linear SVC
    svc = LinearSVC()
    svc.fit(Xtrain, ytrain)
    svc_ypred = svc.predict(Xtest)
    
    #print classification report
    print(classification_report(ytest, svc_ypred)+"\n\n")
    #print Accuracy
    print("AUC Score:- %.3f \n\n" % roc_auc_score(ytest, svc_ypred))
    
    #print confusion matrix
    cnf_mtx(ytest, svc_ypred, "L SVC-ConfusionMatrix.png")
    #plot ROC curve
    plot_roc_curve(svc, Xtest, ytest)
    #axis labels
    plt.xlabel("False Positive Rate")
    plt.ylabel("True Positive Rate")
    #show plot
    plt.show()
    
    
    print("Gaussian NB:- ")
    #GaussianNB
    gnb = GaussianNB()
    gnb.fit(Xtrain, ytrain)
    gnb_ypred = gnb.predict(Xtest)
    
    #print classification report
    print(classification_report(ytest, gnb_ypred)+"\n\n")
    #print accuracy score
    print("AUC Score:- %.3f \n\n" % roc_auc_score(ytest, gnb_ypred))
   
    #print confusion matrix
    cnf_mtx(ytest, gnb_ypred, "GaussionNB-ConfusionMatrix.png")
    #plot ROC curve
    plot_roc_curve(gnb, Xtest, ytest)
    #axis labels
    plt.xlabel("False Positive Rate")
    plt.ylabel("True Positive Rate")
    #show plot
    plt.show()  


    #Beste Performance of ML Model is RandomForestClassifier and GaussianNB
    print("--- Best performance of ML model are RandomForestClassifier and GaussainNB---")
    print("--- As their accuracy score is almost equal ---")
    
    pX_val = pipline.transform(X_val)

    y_val = rnclf.predict(pX_val)
    y1_val = gnb.predict(pX_val)

    pd.DataFrame({'ID_code': test['ID_code'], 'target': y_val}).to_csv("submissionRFC-p.csv", index=False)
    pd.DataFrame({'ID_code': test['ID_code'], 'target':y1_val}).to_csv("submissionGNB-p.csv", index=False)
    
    print("--- End of the program ---")

